File: weather_compare.py
# weather_compare.py
import sys
import json
import requests
from datetime import datetime, timedelta
from collections import OrderedDict

CACHE = {}
CACHE_TTL = 300  # 5 минут

def get_weather(city):
    """Получить данные о погоде для города с кешированием."""
    now = datetime.now()
    if city in CACHE and (now - CACHE[city]['time']).total_seconds() < CACHE_TTL:
        return CACHE[city]['data']
    try:
        url = f"https://wttr.in/{city}?format=j1&lang=ru"
        resp = requests.get(url, timeout=10)
        if resp.status_code != 200:
            return None
        data = resp.json()
        current = data.get('current_condition', [{}])[0]
        weather = {
            'city': city,
            'temp': current.get('temp_C', 'N/A'),
            'humidity': current.get('humidity', 'N/A'),
            'wind': current.get('windSpeed', 'N/A'),
            'pressure': current.get('pressure', 'N/A'),
            'cloud': current.get('cloudcover', 'N/A'),
            'desc': current.get('weatherDesc', [{}])[0].get('value', ''),
        }
        CACHE[city] = {'time': now, 'data': weather}
        return weather
    except Exception as e:
        print(f"Ошибка получения погоды для {city}: {e}", file=sys.stderr)
        return None

def print_table(weather_list):
    """Вывод таблицы с погодой."""
    if not weather_list:
        print("Нет данных для отображения.")
        return
    headers = ['Город', 'Темп.', 'Влаж.', 'Ветер', 'Давл.', 'Облач.', 'Описание']
    col_widths = [15, 8, 8, 8, 8, 8, 20]
    # Печать заголовка
    header_line = '|'.join(h.center(w) for h, w in zip(headers, col_widths))
    print('+' + '+'.join('-'*w for w in col_widths) + '+')
    print('|' + header_line + '|')
    print('+' + '+'.join('-'*w for w in col_widths) + '+')
    for w in weather_list:
        if w is None:
            continue
        row = [
            w['city'][:col_widths[0]].center(col_widths[0]),
            f"{w['temp']}°C".center(col_widths[1]),
            f"{w['humidity']}%".center(col_widths[2]),
            f"{w['wind']} км/ч".center(col_widths[3]),
            f"{w['pressure']} мм".center(col_widths[4]),
            f"{w['cloud']}%".center(col_widths[5]),
            w['desc'][:col_widths[6]].center(col_widths[6])
        ]
        print('|' + '|'.join(row) + '|')
    print('+' + '+'.join('-'*w for w in col_widths) + '+')

def interactive():
    """Интерактивный режим."""
    cities = []
    print("Weather Compare (интерактивный режим)")
    print("Команды: add <город>, remove <город>, show, save, load, quit")
    while True:
        cmd = input("> ").strip().split()
        if not cmd:
            continue
        if cmd[0] == 'quit':
            break
        elif cmd[0] == 'add' and len(cmd) > 1:
            city = ' '.join(cmd[1:])
            if city not in cities:
                cities.append(city)
                print(f"Город {city} добавлен.")
            else:
                print(f"Город {city} уже есть.")
        elif cmd[0] == 'remove' and len(cmd) > 1:
            city = ' '.join(cmd[1:])
            if city in cities:
                cities.remove(city)
                print(f"Город {city} удалён.")
            else:
                print(f"Город {city} не найден.")
        elif cmd[0] == 'show':
            if not cities:
                print("Список городов пуст.")
                continue
            weather_list = []
            for city in cities:
                data = get_weather(city)
                if data:
                    weather_list.append(data)
            print_table(weather_list)
        elif cmd[0] == 'save':
            with open('favorites.txt', 'w', encoding='utf-8') as f:
                f.write('\n'.join(cities))
            print("Список сохранён в favorites.txt")
        elif cmd[0] == 'load':
            try:
                with open('favorites.txt', 'r', encoding='utf-8') as f:
                    loaded = [line.strip() for line in f if line.strip()]
                cities = loaded
                print("Список загружен из favorites.txt")
            except FileNotFoundError:
                print("Файл favorites.txt не найден.")
        else:
            print("Неизвестная команда.")

def main():
    if len(sys.argv) == 1:
        interactive()
        return
    cities = sys.argv[1:]
    weather_list = []
    for city in cities:
        data = get_weather(city)
        if data:
            weather_list.append(data)
    print_table(weather_list)

if __name__ == '__main__':
    main()

File: weather_compare.js
// weather_compare.js
const axios = require('axios');
const fs = require('fs');
const readline = require('readline');

const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout
});

const cache = {};
const CACHE_TTL = 300 * 1000; // milliseconds

async function getWeather(city) {
    const now = Date.now();
    if (cache[city] && (now - cache[city].time < CACHE_TTL)) {
        return cache[city].data;
    }
    try {
        const url = `https://wttr.in/${city}?format=j1&lang=ru`;
        const resp = await axios.get(url, { timeout: 10000 });
        if (resp.status !== 200) throw new Error(`HTTP error: ${resp.status}`);
        const data = resp.data;
        const current = data.current_condition && data.current_condition[0];
        if (!current) throw new Error('No current_condition');
        const weather = {
            city: city,
            temp: current.temp_C || 'N/A',
            humidity: current.humidity || 'N/A',
            wind: current.windSpeed || 'N/A',
            pressure: current.pressure || 'N/A',
            cloud: current.cloudcover || 'N/A',
            desc: (current.weatherDesc && current.weatherDesc[0] && current.weatherDesc[0].value) || ''
        };
        cache[city] = { data: weather, time: now };
        return weather;
    } catch (e) {
        throw new Error(`Ошибка получения погоды для ${city}: ${e.message}`);
    }
}

function center(s, width) {
    if (s.length >= width) return s.substring(0, width);
    const left = Math.floor((width - s.length) / 2);
    const right = width - s.length - left;
    return ' '.repeat(left) + s + ' '.repeat(right);
}

function printSeparator(widths) {
    process.stdout.write('+');
    widths.forEach(w => process.stdout.write('-'.repeat(w) + '+'));
    console.log();
}

function printRow(row, widths) {
    process.stdout.write('|');
    row.forEach((s, i) => process.stdout.write(center(s, widths[i]) + '|'));
    console.log();
}

function printTable(weatherList) {
    if (weatherList.length === 0) {
        console.log('Нет данных для отображения.');
        return;
    }
    const headers = ['Город', 'Темп.', 'Влаж.', 'Ветер', 'Давл.', 'Облач.', 'Описание'];
    const widths = [15, 8, 8, 8, 8, 8, 20];
    printSeparator(widths);
    printRow(headers, widths);
    printSeparator(widths);
    weatherList.forEach(w => {
        const row = [
            center(w.city, widths[0]),
            center(w.temp + '°C', widths[1]),
            center(w.humidity + '%', widths[2]),
            center(w.wind + ' км/ч', widths[3]),
            center(w.pressure + ' мм', widths[4]),
            center(w.cloud + '%', widths[5]),
            center(w.desc, widths[6])
        ];
        printRow(row, widths);
    });
    printSeparator(widths);
}

function interactive() {
    const cities = [];
    console.log('Weather Compare (интерактивный режим)');
    console.log('Команды: add <город>, remove <город>, show, save, load, quit');
    const prompt = () => {
        rl.question('> ', async (line) => {
            const parts = line.trim().split(/\s+/);
            if (parts.length === 0) { prompt(); return; }
            switch (parts[0]) {
                case 'quit':
                    rl.close();
                    return;
                case 'add':
                    if (parts.length < 2) {
                        console.log('Укажите город.');
                        prompt(); return;
                    }
                    const city = line.substring(parts[0].length + 1);
                    if (!cities.includes(city)) {
                        cities.push(city);
                        console.log(`Город ${city} добавлен.`);
                    } else {
                        console.log(`Город ${city} уже есть.`);
                    }
                    prompt();
                    break;
                case 'remove':
                    if (parts.length < 2) {
                        console.log('Укажите город.');
                        prompt(); return;
                    }
                    const cityRem = line.substring(parts[0].length + 1);
                    const idx = cities.indexOf(cityRem);
                    if (idx !== -1) {
                        cities.splice(idx, 1);
                        console.log(`Город ${cityRem} удалён.`);
                    } else {
                        console.log(`Город ${cityRem} не найден.`);
                    }
                    prompt();
                    break;
                case 'show':
                    if (cities.length === 0) {
                        console.log('Список городов пуст.');
                        prompt(); return;
                    }
                    const weatherList = [];
                    for (const c of cities) {
                        try {
                            weatherList.push(await getWeather(c));
                        } catch (e) {
                            console.log(e.message);
                        }
                    }
                    printTable(weatherList);
                    prompt();
                    break;
                case 'save':
                    fs.writeFileSync('favorites.txt', cities.join('\n'));
                    console.log('Список сохранён в favorites.txt');
                    prompt();
                    break;
                case 'load':
                    try {
                        const data = fs.readFileSync('favorites.txt', 'utf-8');
                        const loaded = data.split('\n').map(s => s.trim()).filter(s => s);
                        cities.length = 0;
                        cities.push(...loaded);
                        console.log('Список загружен из favorites.txt');
                    } catch (e) {
                        console.log('Файл favorites.txt не найден.');
                    }
                    prompt();
                    break;
                default:
                    console.log('Неизвестная команда.');
                    prompt();
            }
        });
    };
    prompt();
}

async function main() {
    const args = process.argv.slice(2);
    if (args.length === 0) {
        interactive();
    } else {
        const weatherList = [];
        for (const city of args) {
            try {
                weatherList.push(await getWeather(city));
            } catch (e) {
                console.log(e.message);
            }
        }
        printTable(weatherList);
    }
}

main().catch(console.error);
